### DISCUSSION TITLE
Realtime App Updates

### DISCUSSION THREAD

**Discussion by Fusseldieb** [directus/directus#3221] [Feature Requests]
A thing that I've noticed that could eventually be interesting in the future are Websockets or similar.

I noticed that when Directus is open on multiple computers/screens, if the database is updated or something is inserted, it doesn't get updated on other screens. This obviously happens (or doesn't happen) because the SPA is static. Things get loaded and that's it. But why _stop_ there _(~no~ pun intended)_...

With websockets or similar in place, when something is altered, the changes (or request to reload the collection) would update the table and it would always show the newest and greatest version of the data.

It would then be really _real-time_ 😀

Would love to hear opinions on this idea!

**Comment by rijkvanzanten** [comment]
I'm 100% on board with this 💯 I do think it should be opt-in though, otherwise you'll go crazy with the amount of changes if you're in a very active project. I can see this as a sidebar module that allows you to configure "auto refresh" with options: real-time, 15 seconds, 30 seconds, 5 minutes).

A big part of this is to support `ws://` connections to the REST API in the first place, which is an interesting discussion too.

@Fusseldieb Mind opening a second feature request for `ws://` (and `wss://`) support for the REST endpoints, with ideally somewhat of a spec what the IO should look like? ✨

**Reply by Fusseldieb** [reply]
> I'm 100% on board with this 💯 I do think it should be opt-in though, otherwise you'll go crazy with the amount of changes if you're in a very active project. I can see this as a sidebar module that allows you to configure "auto refresh" with options: real-time, 15 seconds, 30 seconds, 5 minutes).

Good idea.

> A big part of this is to support ws:// connections to the REST API in the first place, which is an interesting discussion too.

I was thinking... while it's interesting to add Websockets to Directus as a whole, supporting every operation/endpoint would be time consuming to implement, and it would need to be implemented in every aspect of Directus, so I would suggest to just add an "overlay" to begin with.
With overlay I mean just adding a single command which then would refresh the currently displayed data, giving Directus a "Real-time" feel without needing to implement ws into everything in the meantime.

Somewhat like this:

```
API/WS->Client: NEW_DATA_AVAILABLE
Client->API/WS: OK
App: Reload/Overwrite currently present data (over HTTP, like it would be done when filtering something)
```

While it wouldn't be perfect, or use the maginally faster speed of loading data *directly* over ws, it would allow this feature to be implemented much faster, since it wouldn't need to be deeply integrated into every "moving" part in Directus.

> @Fusseldieb Mind opening a second feature request for ws:// (and wss://) support for the REST endpoints, with ideally somewhat of a spec what the IO should look like? ✨

No problem, however, I don't feel *qualified* to lay down the IO for Websockets, but I can give a rough overall idea.

**Reply by rijkvanzanten** [reply]
> I was thinking... while it's interesting to add Websockets to Directus as a whole, supporting every operation/endpoint would be time consuming to implement, and it would need to be implemented in every aspect of Directus, so I would suggest to just add an "overlay" to begin with.

This is fair, but there's one massive caveat: we can't change this after it's launched. Removing this temporary overlay down the road in favor of "full" WS support means that it's a breaking change and thus a new major version.. I'm generally more of camp "lets get it right" the first time instead of doing a quick and dirty version first that needs to be swapped out later.

> I don't feel qualified to lay down the IO for Websockets

Me neither! I've never really seen / used a "full fledged" "REST like" websockets API :o

**Reply by Fusseldieb** [reply]
You made a valid point. It would be a breaking change...

> I'm generally more of camp "lets get it right" the first time instead of doing a quick and dirty version first that needs to be swapped out later.

I fully agree with that!

 > Me neither! I've never really seen / used a "full fledged" "REST like" websockets API :o

I'm already writing a extensive feature request on how this could look like. But as I said, since I have never ever laid down something like that, there could be flaws. I would love if you would chime in there as soon as I'm done writing and submitting it :)

**Reply by cynova** [reply]
One option that should be relatively easy to implement:

1. Simply proxy all REST Endpoints over WSS (with a path and method params, maybe socket.io)
2. Add a boolean param "subscribe" which does: return data as normal, determine what collections are contained in output. We can then watch for changes via the activity API (listen for changes via the path that webhooks work with)
3. Once such an event fires we re-evaluate the query and send it over WS if it actually did change (hash?)

**Reply by josias-r** [reply]
Keep GraphQL subscriptions in mind too. I think a potential solution should not just cover REST endpoints, but also GraphQL subscriptions.

**Reply by joeinnes** [reply]
> Me neither! I've never really seen / used a "full fledged" "REST like" websockets API :o

Feathers.js does a great job of this - might be worth looking into :)

**Reply by rijkvanzanten** [reply]
@joeinnes Seems like they're doing everything through a single ws entrypoint, with message names to scope what to do / listen to (like `articles created`, or `patch messages`)

**Reply by benhaynes** [reply]
Just referencing this API Websockets Discussion: https://github.com/directus/directus/discussions/3255

**Comment by liho00** [comment]
https://next-event-coverage-liveblog.vercel.app/  <-- here is an excellent example of realtime app update with GraphQL event streaming. 

Websocket personally thinking is not quite good for realtime update, due to its does not have reconnect mechanism, need to self implement.

Not sure about graphql event streaming, i need to futher research on it. But, looking good so far.

![image](https://user-images.githubusercontent.com/77238199/144215828-919306db-2684-430c-b2e9-c74147d6e8bf.png)

**Reply by Fusseldieb** [reply]
SSE also has some drawbacks. 
I've read that SSE is kinda a subscribe-receive thing (only UTF-8, no binary) where as WebSockets are bidirectional and _also_ binary. I think this needs to be discussed. 

https://stackoverflow.com/questions/5195452/websockets-vs-server-sent-events-eventsource

**Reply by rijkvanzanten** [reply]
@Fusseldieb Absolutely agree. Websockets can achieve the same functionality as SSE, without its ceilings. To me, the two-way binding of websockets is where the real magic of real time communication comes in.

**Reply by liho00** [reply]
It looks like websocket would be better haha

**Reply by eliteproxy7** [reply]
I found this pubic repo, worked like a charm for me!

https://github.com/br41nslug/directus-websocket-subscribe

**Reply by rijkvanzanten** [reply]
@br41nslug did a great job in this custom extension implementation. Can't wait to keep discussing and imagining what a native /webhooks endpoint would look like!

On that note, I was even thinking that instead of relying on a single endpoint, we could also allow for using `ws://` on the regular REST API endpoints, and handle the subscription scope that way :o

**Reply by liho00** [reply]
@rijkvanzanten  I found 2 ways of the subscription through websocket, based on param or querystring & based on message payload, we also possible to doing authorization via websocket (first screenshot)

> On that note, I was even thinking that instead of relying on a single endpoint, we could also allow for using ws:// on the regular REST API endpoints, and handle the subscription scope that way :o

1. based on message payload
<img width="1239" alt="quick_start_message-367c9f516d7cc6803a1075be28469bf8" src="https://user-images.githubusercontent.com/77238199/144696723-9c6f56bf-1ebd-4c68-8c69-640cf8a44889.png">

2.  based on param or querystring
![image](https://user-images.githubusercontent.com/77238199/144696817-a6254658-7173-4f64-a5b7-f8c890aa6406.png)

**Reply by eliteproxy7** [reply]
> @br41nslug did a great job in this custom extension implementation. Can't wait to keep discussing and imagining what a native /webhooks endpoint would look like!
> 
> On that note, I was even thinking that instead of relying on a single endpoint, we could also allow for using `ws://` on the regular REST API endpoints, and handle the subscription scope that way :o

we can look at it in the same flow as the webhooks. The user would create a real-time service for a collection etc and create the endpoint they can subscribe to

**Reply by CallMeLaNN** [reply]
I just looking for graphql subscription...

Consider client libraries for REST API other than the sdk during the design. I think mostly use an endpoint with namespace/room etc... "they all have different opinions on how real-time should be done": *Primus* try to make it common and support socket.io, sockjs, WebSockets etc. but another way is to let admin choose a ws hook implementation as a separate module to opt-in.

Anyway supporting graphql subscription is the way to go first since front-end dev can use existing graphql client.

As for many vs single endpoint. I open to both but prefer single for reduced concurrent connections and possibility to fallback to single long pooling. Use case on client: one place for any notifications and badge number for busy collections. I think activities tracking too many info.

Subscribe to specific collection (and reuse filter like _eq) useful too. E.g. subscribe only comments collection for a particular post_id. Graphql support this using withFilter.

**Comment by rijkvanzanten** [comment]
Next to websockets, also worth looking into MQTT support (needs more research, 80/20): https://github.com/directus/directus/discussions/14309

**Comment by onurbamaro** [comment]
There is a community extension (that I have not tested). 

Worth looking: https://github.com/erdemozveren/directus-realtime-extension

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by connorwinston** [comment]
This was implemented in version 10.3!

**Reply by rijkvanzanten** [reply]
Not so quick Connor! WebSockets support in the API was implemented, but the app isn't using it yet. That's what this feature request is about 🙂

**Comment by w0kyj** [comment]
Break down of needed components
- collaborative editing (look for existing RFC/discussions)
    - notice someone else is editing an item
    - someone else saves the item -- get notified
- auto refresh currently supported, but uses polling -- change this to websockets
- auto save
    - interval for save (configurable?)
    - benefit to new versioning and live preview for true realtime
- data model update --> auto refresh
    - refresh user sessions when data model changes saved
